corehq/apps/cloudcare/tests/test_utils.py

```diff
@@ -107,3 +119,103 @@ def setUpClass(cls):
+
+    def test_commcare_user_has_access_if_assigned_role_that_can_access_all_web_apps(self):
+        role = UserRole.create(
+            self.domain, "can-access-all-web-apps", permissions=HqPermissions(access_web_apps=True)
+        )
+        self.commcare_user.set_role(self.domain, role.get_qualified_id())
+
+        has_access = can_user_access_web_app(self.app_doc, self.commcare_user, self.domain)
+
+        self.assertTrue(has_access)
+
+    def test_commcare_user_has_access_if_assigned_role_that_cannot_access_web_apps(self):
+        # mobile users have not historically required this new permission, so we need to assume they have access
+        role = UserRole.create(
+            self.domain, "cannot-access-web-apps", permissions=HqPermissions(access_web_apps=False)
+        )
+        self.commcare_user.set_role(self.domain, role.get_qualified_id())
+
+        can_access = can_user_access_web_app(self.app_doc, self.commcare_user, self.domain)
+
+        self.assertTrue(can_access)
+
+    def test_commcare_user_has_access_if_assigned_role_that_can_access_specific_app(self):
+        role = UserRole.create(
+            self.domain,
+            "can-access-specific-app",
+            permissions=HqPermissions(web_apps_list=[self.app_doc["copy_of"]]),
+        )
+        self.commcare_user.set_role(self.domain, role.get_qualified_id())
+
+        has_access = can_user_access_web_app(self.app_doc, self.commcare_user, self.domain)
+
+        self.assertTrue(has_access)
+
+    def test_web_user_has_access_if_assigned_role_that_can_access_all_web_apps(self):
+        role = UserRole.create(
+            self.domain, "can-access-all-web-apps", permissions=HqPermissions(access_web_apps=True)
+        )
+        self.web_user.set_role(self.domain, role.get_qualified_id())
+
+        has_access = can_user_access_web_app(self.app_doc, self.web_user, self.domain)
+
+        self.assertTrue(has_access)
+
+    def test_web_user_does_not_have_access_if_assigned_role_that_cannot_access_web_apps(self):
+        role = UserRole.create(
+            self.domain, "cannot-access-web-apps", permissions=HqPermissions(access_web_apps=False)
+        )
+        self.web_user.set_role(self.domain, role.get_qualified_id())
+
+        has_access = can_user_access_web_app(self.app_doc, self.web_user, self.domain)
+
+        self.assertFalse(has_access)
+
+    def test_web_user_has_access_if_assigned_role_that_can_access_specific_app(self):
+        role = UserRole.create(
+            self.domain,
+            "can-access-specific-app",
+            permissions=HqPermissions(web_apps_list=[self.app_doc["copy_of"]]),
+        )
+        self.web_user.set_role(self.domain, role.get_qualified_id())
+
+        has_access = can_user_access_web_app(self.app_doc, self.web_user, self.domain)
+
+        self.assertTrue(has_access)
+
+    @flag_enabled("WEB_APPS_PERMISSIONS_VIA_GROUPS")
+    def test_commcare_user_has_access_if_in_group(self):
```

Comment:
Could you add a negative test case?

Reply:
Also a test case for a web user on a domain that uses this flag?

Reply:
It's hypothetically possible for permissions to grant access to *different* apps than groups, which would also be good to test because it's not immediately obvious what the behavior *should* be. From the description of https://github.com/dimagi/commcare-hq/pull/34421:

> In addition to [the new app-specific changes], any legacy ApplicationAccess restrictions will continue to be applied, for backwards compatibility.

Reply:
Between 92ffbbc2c5c2e55fa3d713b6cc2b891c67413c02 and a995265e9b2288d0e36e868bd1d75de37a0b00e7 I _think_ I've touched on all of these needed test cases.